### PR TITLE
Add table potential class to forcefields library

### PR DESCRIPTION
**Pull request by chrisjonesBSU**
This PR adds a class that creates hoomd table potentials. In hoomd, the table potentials are just created from arrays (i.e. you don't give hoomd the txt or npy file). So, this class is designed to accept dictionaries of `{type: {}}` where the inside dictionary has keys and values of `{parameter: parameter_value}` for all of the hoomd parameters (force, energy, etc...).  However, since the use case is most commonly going to be loading table potential files, I added a `from_files` class method that handles building up of the these dictionaries.

To Do:

- [x] Doc strings
- [x] Unit tests
- [ ] More type checks and errors/warnings

### CODE REVIEW

File: hoomd_organics/library/forcefields.py [hunk 1]
```diff
@@ -187,3 +188,204 @@ def _create_forcefield(self):
...
+        nlist_buffer=0.40,
+    ):
+        """Create forcefield from text files containing tabulated forces."""
+        # Read pair files
+        pair_dict = dict()
+        r_min = set()
+        r_max = set()
+        if pairs:
+            for pair_type in pairs:
+                table = np.loadtxt(pairs[pair_type])
```
Comment (chrisjonesBSU) [inline hoomd_organics/library/forcefields.py:235]: Need a way to handle `npy` files

File: hoomd_organics/library/forcefields.py [hunk 2]
```diff
@@ -187,3 +188,204 @@ def _create_forcefield(self):
...
+                r = table[:, 0]
+                r_min.add(r[0])
+                r_max.add(r[-1])
+                pair_dict[pair_type] = dict()
+                pair_dict[pair_type]["U"] = table[:, 1]
+                pair_dict[pair_type]["F"] = table[:, 2]
+            if len(r_min) != len(r_max) != 1:
```
Comment (chrisjonesBSU) [inline hoomd_organics/library/forcefields.py:242]: Technically, we can handle pair tables with different r_mins and r_maxs if we set the rmax on a per-pair basis in `_create_hoomd_forcefield`

File: hoomd_organics/library/forcefields.py [hunk 3]
```diff
@@ -187,3 +188,204 @@ def _create_forcefield(self):
...
+        if dihedrals:
+            for dih_type in dihedrals:
+                table = np.loadtxt(dihedrals[dih_type])
+                thetas = table[:, 0]
+                if thetas[0] != 0 or not np.allclose(
+                    thetas[-1], 2 * np.pi, atol=1e-5
+                ):
+                    raise ValueError(
+                        "Dihedral angle values must be evenly spaced and "
+                        "range from -Pi to Pi."
```
Comment (chrisjonesBSU) [inline hoomd_organics/library/forcefields.py:286]: Need to double check this

File: hoomd_organics/library/forcefields.py [hunk 4]
```diff
@@ -187,3 +188,272 @@ def _create_forcefield(self):
...
+                    ("A", "A"): "A_pairs.txt
+                    ("B", "B"): "B_pairs.txt
+                    ("A", "B"): "AB_pairs.txt
+                },
+                bonds = {"A-A": "A_bonds.txt", "B-B": "B_bonds.txt"},
+                angles = {"A-A-A": "A_angles.txt", "B-B-B": "B_angles.txt"},
+            )
+        """
+
+        def _load_file(file):
```
Comment (chrisjonesBSU) [inline hoomd_organics/library/forcefields.py:290]: I wonder if we need a way to pass in np kwargs here by adding a kwargs param to the class.

Comment (marjanAlbouye) [reply hoomd_organics/library/forcefields.py:290]: what do you mean by np kwargs?

File: hoomd_organics/library/forcefields.py [hunk 5]
```diff
@@ -182,3 +183,284 @@ def _create_forcefield(self):
...
+            elif file.split(".")[-1] == "npy":
+                return np.load(file, **kwargs)
+
+        # Read pair files
+        pair_dict = dict()
+        pair_r_min = set()
+        pair_r_max = set()
+        if pairs:
+            for pair_type in pairs:
+                table = _load_file(pairs[pair_type], **kwargs)
```
Comment (marjanAlbouye) [inline hoomd_organics/library/forcefields.py:310]: Once we load a file we can have a set of validations/checks on the loaded files to make sure that the inputs follow the expected format. For example, we should make sure that loaded table has 3 columns. Also making sure that `r` is  an array and has two values.

File: hoomd_organics/library/forcefields.py [hunk 6]
```diff
@@ -182,3 +183,284 @@ def _create_forcefield(self):
...
+        It is assumed that the structure of the files are:
+            * Column 1: Independent variable (e.g. distance, length, angle)
+            * Column 2: Energy
+            * Column 3: Force
+
+        """
+
+        def _load_file(file, **kwargs):
+            """Call the correct numpy method."""
+            if file.split(".")[-1] in ["txt", "csv"]:
```
Comment (marjanAlbouye) [inline hoomd_organics/library/forcefields.py:299]: Also probably need some validations here on the file name and file path. 
For example, raising error if the file is not txt or csv or numpy or doing a `os.path.exists` check before loading.

**Review by marjanAlbouye — APPROVED** [review]
Everything looks good. I just added a few comments mostly around checks and validations when loading the files, which I think you already have those in mind.
Thank you

File: hoomd_organics/library/forcefields.py [hunk 7]
```diff
@@ -182,3 +183,284 @@ def _create_forcefield(self):
...
+    It may be most convenient to store tabulated data in files,
+    in that case use the `from_files` method.
+
+    Parameters
+    ----------
+    pairs: dict, optional, default None
+    bonds: dict, optional, default None
+    angles: dict, optional, default None
+    dihedrals: dict, optional, default None
+
```
[the remark this answers is not in the document]
Comment (chrisjonesBSU) [reply hoomd_organics/library/forcefields.py:214]: Yeah, I forgot that I need to finish this part of the doc strings still.

File: hoomd_organics/library/forcefields.py (hunk 6 above)
Comment (chrisjonesBSU) [reply hoomd_organics/library/forcefields.py:299]: Good idea.

File: hoomd_organics/library/forcefields.py (hunk 5 above)
Comment (marjanAlbouye) [reply hoomd_organics/library/forcefields.py:310]: Oh, nvm! I was looking at the way you get `pair_r_min` and `pair_r_max` and I thought they are both coming from one row of the table. Checking  for a minimum row numbers of two sounds reasonable.

Comment (chrisjonesBSU) [reply hoomd_organics/library/forcefields.py:310]: Ok, I see. They are pulling the first and last values of the r column, which should be r_min and r_max (i.e. r_cut)

**Pull request merged**